tket2/src/rewrite_space/factory/kak.rs

```diff
@@ -0,0 +1,255 @@
+    }
+}
+
+pub struct KAKFactory;
+
+struct SearchingWalker <'w> {
+    walker: Walker <'w>,                    // Initial state of walker.
+    frontier_wires: Vec<PersistentWire>,    // Wires to be expanded.
+    pattern: KAKResynthesis,                // Current pattern.
+    two_qubit_node: Option<PatchNode>,      // If a 2q gate has been found along one wire,
+                                            // but not yet the other, then it is stored here.
+}
+
+impl CommitFactory for KAKFactory{
+    type PatternMatch = KAKResynthesis;
+
+    type Cost = usize;
+
+    const PATTERN_RADIUS: usize = 5;
+
+    fn find_pattern_matches<'w: 'a, 'a>(
+        &'a self,
+        pattern_root: PatchNode,
+        walker: Walker<'w>,
+    ) -> impl Iterator<Item = (Self::PatternMatch, Walker<'w>)> + 'a {
+
+        // A list of matches to be returned.
+        let mut matches: Vec<(Self::PatternMatch, Walker<'w>)> = Vec::new();
+        
+        // If the node is not a CZ gate, the matches are empty.
+        if walker.as_hugr_view().get_optype(pattern_root) != &Tk2Op::CZ.into() {
+            return matches.into_iter();
+        }
+
+        // A list of all the searching walkers.
+        let mut searching_walker_list = Vec::new();
+        // Create an initial searching walker, starting at the CZ gate discovered.
+        searching_walker_list.push(
+            SearchingWalker{
+                walker: walker,
+                frontier_wires: vec![
+                    walker.get_wire(pattern_root, OutgoingPort::from(0)),
+                    walker.get_wire(pattern_root, OutgoingPort::from(1)),
+                ],
+                pattern: KAKResynthesis { wires: Vec::new() },
+                two_qubit_node: None,
+            }
+        );
+
+        // Continue for as long as there are searching walkers to explore.
+        while let Some(mut searching_walker) = searching_walker_list.pop() {
+            
+            // If there are no wires to be searched then the match gets pushed
+            // to the list of matches.
+            let Some(wire_to_search) = searching_walker.frontier_wires.pop() else {
+                matches.push((searching_walker.pattern, searching_walker.walker));
+                continue;
+            };
+            
+            // Assuming now that there are wires to be searched, get
+            // the input port of the wire. There should be exactly one.
+            let (frontier_node, _) = searching_walker.walker.wire_pinned_inports(&wire_to_search).next().expect("error"); 
+            
+            // Expand the wire into the future
+            for expanded_walker in searching_walker.walker.expand(&wire_to_search, Direction::Outgoing) {
+                
+                // Get the new expanded version of the wire to search along.
+                let expanded_wire = expanded_walker.get_wire(
+                    frontier_node, 
+                    IncomingPort::from(0),
```

Comment:
I don't think this will always be port 0? Also, if you change the above direction, you will need to change it here too.